Tasks in comments support

stilliard opened the pull request:
Looking to support PR comments & PR review comments with tasks inside.

Fixes #2

Previewing via PR but all goes well I will merge next week when I'm back :)

hitmands commented:
:rocket:

hjkatz commented on index.js:
```diff
@@ -1,21 +1,57 @@
+    let hasOutstandingTasks = checkOutstandingTasks(pr.body);
+
+    // lookup comments on the PR
+    let comments = await context.github.issues.listComments(context.repo({
```
What happens if these requests fail?

stilliard replied:
Thanks @hjkatz great point! The requests only use Github's api but you're right they could certainly fail. 
Currently what will happen is it'll error, loudly in glitch for the logs. 
Thinking more about this, it could even error I guess if you don't have the new permissions granted. 
I think then the safest option is if these fail/error then we skip that check?

hjkatz replied:
IMO I think the opposite state would be desired, where the check is by default a failure/pending until the API works. This of course could be a problem when the API calls fail with bad permissions or with a bug in the code.

It's up to you though :)

stilliard replied:
It's a tricky one, if someone's using this and they haven't granted the new permissions we need, I still want the app to work for them. Also if Github's having some issues, I'd rather not hold up releases, though I guess in that case it could be better to temp disable it as a required check when needed. 

I'll have a look through github's api docs and see if we can easily tell the difference between permission issues & actual api errors as then we can have best of both :)

billsacks replied:
For what it's worth, I prefer the conservative approach @hjkatz suggests: If the app can't determine that things are successful (even if due to permissions issues) then it should report a failure. I'm not sure what other GitHub apps do in this situation, though.

stilliard replied:
I agree with api errors, but I disagree about permissions though, a lot of people have already installed the app, this new PR asks for new permissions and if someone has already installed the app and does not want to accept the new permissions for whatever reason I would not want to force them to by breaking their app experience.

billsacks replied:
Will this app do anything useful if they don't accept the permissions? My thinking is that, if the app is just going to always return a "pass" result in this case, then this isn't very useful and could be harmful in that it's misleading; in this case, they'd be better off uninstalling the app. But if the app can still do some useful work without these permissions, then I see your point.

billsacks replied:
(But I'll also reiterate that I don't have much experience with GitHub apps in general, and particularly with maintaining backwards compatibility of them. I certainly see the rationale of not wanting to break things for existing users. So I don't want to argue my points too strongly, I just wanted to let you know my thinking on this.)

stilliard replied:
I think you're thinking of new users, they wouldn't be able to install the app anyway without accepting all new permissions. 
But, existing users, they would want the app to continue working. 

The flow I believe is when we update our app permissions, existing users get an email to say our app is requesting new permissions and can click to accept them. However, they might not see this for a while, or may choose not to give the new permissions for whatever reason. 
Without the new permission, we can continue to work as we did before, what I'm thinking is the new api request that looks at the comments that requires the new permission can likely be wrapped in a try catch or detect otherwise that it failed (will investigate this workflow shortly) and will just continue if there's a permission error. 

This then allows new users to sign up with full new permissions, while allowing the existing app to continue working for existing installs the way it did before untill they accept the new permissions giving the app the new capabilities. 

Hope this clears this up.

billsacks replied:
Okay, thanks for the reply. If it can be made to work the way you describe, then I agree that sounds like good behavior. Thanks!

stilliard replied:
Back to it again now, found time between other projects now I'm back from vacation :)

Had a quick test and looks like actually the new issues permission I've added is only needed for the issue_comment event to be hooked up, checking `context.github.issues` appears to work with and without the new permission meaning the permission not being granted will only not recieve the new issue comment events, no errors to be caught the app side :+1: 

In terms of other possible errors, it's possible if github is having an issue the app could error, but the logs will catch those and from my testing the app doesn't crash, only errors loudly in the logs. 
For now I think that might actually be enough as what will happen is the check will fail to complete and not pass. 

Future improvement I think would be for it to have retry logic? But we can add that in a new PR I think as it'd be good to add this generically for all the api requests we make rather than just these new ones :)

billsacks commented:
Thank you for adding this functionality!

I wanted to check on one aspect: Although I'm not familiar with the GitHub API that is used here, in the python GitHub API, there are four separate places that need to be checked:

1. The overall PR body

2. "Issue" comments

3. PR line comments made in PR reviews

4. The body of PR reviews – i.e., the overall comment that is made during a PR review (not all PR reviews have a body, but some do, if an overall comment is made) (for example, see https://github.com/ESMCI/github-tools/pull/1#pullrequestreview-399407911 - the comment beginning, "In addition to my line comments, please also:")

From what I can see of the changes here, you check (1), (2) and (3), but I don't see an explicit check of (4). Does that need to be added, or in this API is that included in one of the other categories?

stilliard commented:
Hi @billsacks 
Thanks great comments! 
You're right I think I am missing point 4 there. 
I'll test this shortly & get back to you on this thanks again for the review!

stilliard commented:
Hey :wave: 

Thanks @billsacks for finding I wasn't checking the review comments themselves, only the issue comments & diff comments, that's not been fixed to check these as well. 

& thanks to @hjkatz for the review as well prompting more testing for the error handling. 

Looking to merge & release this tomorrow thanks :rocket:

billsacks commented:
Sounds great! Thanks for getting the review comments and for checking out the error handling. The behavior you report sounds reasonable enough to me. Finally, thanks again for your work on this great app!

🎉

stilliard commented:
Just a heads up on this. 

I merged it earlier but the requests to listComments() are timing out on glitch for some reason. 
Running locally it works so I'm wondering if it's an issue with the env on glitch. 

Going to do some more testing later, for now I've commented out the change so it's working as it did before. 

Noticed the checks are also coming back slow today, again locally they are fast so maybe glitch's servers are just running a little slow today. 

Will keep everyone posted on when this is fully released.

stilliard commented:
Tested the live app through my own laptop and it worked fine as well. 
Emailed Glitch to see if they can spot anything, maybe a firewall issue or resource limitation. 
If that's the case I can move the app to my own servers though I do like Glitch's service so we'll see if they can help.

stilliard commented:
This PR is now live :tada: 

No word back from Glitch yet, so for the time being it's self hosted on my own servers & seems to be running faster already.

billsacks commented:
I'm not sure if this is due to this set of changes or some other issue, but I installed this yesterday in a test repository and the app has been running for 17 hours straight now: https://github.com/billsacks/testrepo/pull/5/checks?check_run_id=1000544136 . It finished quickly on a PR with no todo items (https://github.com/billsacks/testrepo/pull/7), but seems to be hanging if there are todo items.

stilliard commented:
Hey @billsacks 
Try retrying the check or edit the pr to trigger it to retry. 
Likely this was during the time moving from glitch to my own server but my checks seem to be completing fine atm.

billsacks commented:
I just tried opening a new PR; it still seems to be hanging (a few minutes in): https://github.com/billsacks/testrepo/pull/8/checks?check_run_id=1008991409

stilliard commented:
Ohh sorry I misunderstood. 
Yes this is due to how we use the "in progress" state rather than completed for when the checklist is still needed to be completed.
Same way some other github checks work with like the [WIP check](https://github.com/apps/wip). 

This way it gives a yellow/orange in progress indicator until the task lists are all completed and then turns green. 

You'll notice it also shows "0 / 1 tasks completed" in the summary title, added that in just after this PR.

stilliard commented:
Hi @billsacks 
Hope you don't mind, I added a comment to your test repo you sent over to check it's working & yep the check summary updated to "0 / 2 tasks completed" when I added a task, and then now shows as "1 / 2 tasks completed" as I've checked off that task I added. 
Hope this is ok.

billsacks commented:
Ah, got it, thanks for clarifying and sorry for my confusion!

stilliard commented:
No worries, great to have you putting it through all this testing! Thank you :+1:
